webapp/src/js/enketo/medic-xpath-extensions.js

```diff
@@ -163,24 +163,28 @@ const dateDiff = function (startDateObj, endDateObj, key) {
   return XPR.number(endDate.diff(startDate, key));
 };
 
-const luhn = function (number) {
+const luhn = function (number, expLength) {
   number = getValue(number);
   number = number.toString().trim().replace(/\s/g, '');
 
-  if (number.length !== 13 || !/^\d+$/.test(number)) {
+  if (!/^\d+$/.test(number) || (expLength && number.length !== expLength)) {
```

Comment:
I think this would be more reusable if the length validation was done as a separate constraint, eg: `cht:validate-luhn(.) and string-length(.) === 13`.

This gives more flexibility in the form construction, for example, you might want `<=` instead of `===`.

What do you think?

Reply:
I've, albeit with limited exposure, seldomly seen the luhn algorithm used without some sort of length validation when it comes to the real world applications as mentioned in the PR description.
That being said, you have a valid point that the type of length check required might differ from impl to impl (`===` vs `<=`).
Still, having the length check outside would mean that the user would also need to strip spaces themselves.
For example `6505  2851  2508 6` and `  6505285125086  ` would satisfy the luhn check as it is written now, but would fail the proposed `cht:validate-luhn(.) and string-length(.) === 13` check in the form. Unless the `string-length` automatically removes all spaces (`.replace(/\s/g, '')`) ?
Thinking about it now, perhaps stripping spaces would also be a useful utility function.
What do you think?

Reply:
Yeah that's a good point.

In the implementation you've made it optional so maybe that's the way to go - it's an optional parameter and you can use it to validate the trimmed length, OR you can omit it if you want to do your own length validation.

If you agree with this optional approach, could you add a test or two that proves it works when you don't provide a length?

Reply:
You're right on the money there! Awesome, I'll get those tests in for you ASAP! 🚀 
Thank you for taking the time to talk through the various approaches; it really is appreciated.